### PR TITLE
cugraph examples

### LINKED ISSUE #2676: Need basic C++ "hello world" example(s) for libcugraph
This request has been made multiple times, but I couldn't find an issue for it so I'm creating this.

Users have requested (a recent hackathon was one know example of such a request) a simple "hello world" style C++ example that uses libcugraph.  For now, we've been recommending the [C++ unit tests](https://github.com/rapidsai/cugraph/tree/branch-22.10/cpp/tests), but these aren't optimized for demonstrating the API as much as they are just achieving test coverage.

One option could be to simply change the unit tests so they demonstrate the API better, or add additional tests that are better suited as examples and leave the other, harder-to-read tests in place for covering edge cases and other test scenarios. This is nice because it guarantees the examples work with the latest code changes since they'll be run as part of CI.

cc @ogreen

### PR DESCRIPTION
**Pull request by naimnv**
This PR adds example on how to use libcugraph to run different algorithms. 

Closes #2676

### CODE REVIEW

File: cpp/examples/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# libcugraph examples
+
+Example codes on how to use libcugraph.
+
+## Contents
+
+- sg_examples: simple code on how to use libcugraph to run different graph algorithms in a signle GPU node.
```
Comment (ChuckHastings) [inline cpp/examples/README.md:7]:
```suggestion
- sg_examples: simple code on how to use libcugraph to run different graph algorithms in a single GPU node.
```

File: cpp/examples/graph_partitioning/graph_partition_examples.cpp [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+  // Meta of the non-owning view of the graph object store vertex/edge partitioning map
+  auto graph_view = graph.view();
+
+  // Total number of vertices
+  vertex_t global_number_of_vertices = graph_view.number_of_vertices();
+
+  // Number of vertices mapped to this process, ie the size of
+  // the vertex partition assigned to this process
+
+  vertex_t size_of_the_vetex_partition_assigned_to_this_process =
```
Comment (ChuckHastings) [inline cpp/examples/graph_partitioning/graph_partition_examples.cpp:125]:
```suggestion
  vertex_t size_of_the_vertex_partition_assigned_to_this_process =
```

File: cpp/examples/graph_partitioning/graph_partition_examples.cpp [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    graph_view.local_vertex_partition_range_size();
+
+  // NOTE: The `renumber_map` contains the vertices assigned to this process.
+
+  // Print verties mapped to this process
+  RAFT_CUDA_TRY(cudaDeviceSynchronize());
+  if (renumber_map) {
+    auto vertex_paritition_title =
```
Comment (ChuckHastings) [inline cpp/examples/graph_partitioning/graph_partition_examples.cpp:133]:
```suggestion
    auto vertex_partition_title =
```

File: cpp/examples/graph_partitioning/graph_partition_examples.cpp [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+      std::string("vertices@rank_").append(std::to_string(comm_rank)).c_str();
+    raft::print_device_vector(
+      vertex_paritition_title, (*renumber_map).data(), (*renumber_map).size(), std::cout);
```
Comment (ChuckHastings) [inline cpp/examples/graph_partitioning/graph_partition_examples.cpp:136]:
```suggestion
      vertex_partition_title, (*renumber_map).data(), (*renumber_map).size(), std::cout);
```

File: cpp/examples/graph_partitioning/graph_partition_examples.cpp [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+  std::vector<vertex_t> h_vertices_in_this_proces((*renumber_map).size());
+
+  raft::update_host(h_vertices_in_this_proces.data(),
+                    (*renumber_map).data(),
+                    (*renumber_map).size(),
+                    handle.get_stream());
+  handle.sync_stream();
+
+  assert(size_of_the_vetex_partition_assigned_to_this_process == (*renumber_map).size());
```
Comment (ChuckHastings) [inline cpp/examples/graph_partitioning/graph_partition_examples.cpp:147]:
```suggestion
  assert(size_of_the_vertex_partition_assigned_to_this_process == (*renumber_map).size());
```

File: cpp/examples/graph_partitioning/graph_partition_examples.cpp [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+    auto number_of_edges = edge_partition.number_of_edges();
+    auto offsets         = edge_partition.offsets();
+    auto indices         = edge_partition.indices();
+
+    auto offsets_title = std::string("offsets_")
+                           .append(std::to_string(comm_rank))
+                           .append("_")
+                           .append(std::to_string(ep_idx))
+                           .c_str();
```
Comment (ChuckHastings) [inline cpp/examples/graph_partitioning/graph_partition_examples.cpp:195]: Can you do this?  Doesn't the string go out of scope once the line is completed?

I feel like you need to leave off the `c_str` call here, and add it to line 197 where you use it in order to guarantee that the memory is not reused.

File: cpp/examples/graph_partitioning/graph_partition_examples.cpp [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+    RAFT_CUDA_TRY(cudaDeviceSynchronize());
+    raft::print_device_vector(offsets_title, offsets.begin(), offsets.size(), std::cout);
+
+    auto indices_title = std::string("indices_")
+                           .append(std::to_string(comm_rank))
+                           .append("_")
+                           .append(std::to_string(ep_idx))
+                           .c_str();
```
Comment (ChuckHastings) [inline cpp/examples/graph_partitioning/graph_partition_examples.cpp:203]: Same comment as above

File: cpp/examples/graph_partitioning/graph_partition_examples.cpp [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+      auto edge_counts  = edge_weight_view->edge_counts();
+
+      assert(number_of_edges == edge_counts[ep_idx]);
+
+      RAFT_CUDA_TRY(cudaDeviceSynchronize());
+      auto weights_title = std::string("weights_")
+                             .append(std::to_string(comm_rank))
+                             .append("_")
+                             .append(std::to_string(ep_idx))
+                             .c_str();
```
Comment (ChuckHastings) [inline cpp/examples/graph_partitioning/graph_partition_examples.cpp:219]: Same here

File: cpp/examples/multi_gpu/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,34 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
```
Comment (ChuckHastings) [inline cpp/examples/multi_gpu/CMakeLists.txt:1]: Don't we need the full apache license block here.

File: cpp/examples/multi_gpu/mg_examples.cpp [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+    cugraph::louvain(handle,
+                     std::optional<std::reference_wrapper<raft::random::RngState>>{std::nullopt},
+                     graph_view,
+                     edge_weight_view,
+                     cluster_assignments.data(),
+                     max_level,
+                     threshold,
+                     resolution);
+
+  std::cout << "rank : " << comm_rank << ", modularity : " << modularity << std::endl;
```
Comment (ChuckHastings) [inline cpp/examples/multi_gpu/mg_examples.cpp:154]: Maybe also display the cluster assignments with `raft::print_device_vector`?

Comment (ChuckHastings) [reply cpp/examples/multi_gpu/mg_examples.cpp:154]: Perhaps add `std::min(cluster_assignments.size(), size_t{20})` as the size in the `print_device_vector` call to demonstrate that you can call it on a subset of the device vector... in case they run a larger graph through the test code.

File: cpp/examples/single_gpu/sg_examples.cpp [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    cugraph::louvain(handle,
+                     std::optional<std::reference_wrapper<raft::random::RngState>>{std::nullopt},
+                     graph_view,
+                     edge_weight_view,
+                     cluster_assignments.data(),
+                     max_level,
+                     threshold,
+                     resolution);
+
+  std::cout << "modularity : " << modularity << std::endl;
```
Comment (ChuckHastings) [inline cpp/examples/single_gpu/sg_examples.cpp:114]: Same thought on adding a `raft::print_device_vector`.

**Comment by seunghwak** [comment]
@naimnv Is this PR ready-for-review?

**Comment by naimnv** [comment]
@seunghwak Yes. I will add an addition function in graph partitioning example, but getting review as of it now is always helpful.

File: cpp/examples/README.md [hunk 12]
```diff
@@ -0,0 +1,32 @@
+# libcugraph examples
+
+Example codes on how to use libcugraph.
+
+## Contents
+
+- single_gpu: simple code on how to use libcugraph to run different graph algorithms in a single GPU node.
+
+- multi_gpu: example code on how to use libcugraph to run different graph algorithms in a multi GPU node.
```
Comment (seunghwak) [inline cpp/examples/README.md:9]: We can run different algorithms using more than one node.

File: cpp/examples/README.md [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+-  graph_partitioning: contains code to explain vertex and edge partitioning in cugraph.
```
Comment (seunghwak) [inline cpp/examples/README.md:11]: What's the purpose of this? Users don't need to understand partitioning to use libcugraph.

Comment (naimnv) [reply cpp/examples/README.md]: For application developer who wants to contribute and for those who want to implement their own/custom algorithms would like to understand how partitioning work.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+                                          handle.get_stream());
+
+    per_v_transform_reduce_incoming_e(
+      handle,
+      graph_view,
+      src_vertex_weights_cache.view(),
+      dst_vertex_weights_cache.view(),
+      (*edge_weight_view),
+      [] __device__(auto src, auto dst, auto src_prop, auto dst_prop, auto edge_prop) {
+        printf("\nsrc ---> %d dst = %d :  src_prop = %f dst_prop = %f edge_prop = %f\n",
```
Comment (seunghwak) [inline cpp/examples/advanced/cugraph_operations/graph_operations.cu:166]: What is "\nsrc ---> %d dst = %d"? Did you intend "src = %d ---> dst = %d"? And why "\n" in both the beginning and the end?

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+
+    cugraph::edge_dst_property_t<graph_view_t, result_t> dst_vertex_weights_cache(handle,
+                                                                                  graph_view);
+
+    update_edge_src_property(handle, graph_view, vertex_weights.begin(), src_vertex_weights_cache);
+
+    update_edge_dst_property(handle, graph_view, vertex_weights.begin(), dst_vertex_weights_cache);
+
+    rmm::device_uvector<result_t> outputs(size_of_the_vertex_partition_assigned_to_this_process,
+                                          handle.get_stream());
```
Comment (seunghwak) [inline cpp/examples/advanced/cugraph_operations/graph_operations.cu:157]: I think more descriptive name (or at least a comment saying what this variable stores considering that this is for novice users) will be helpful, you are summing each vertex's destination neighbor's out-weight_sum * edge_weight, right?

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu (hunk 14 above)
Comment (seunghwak) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu:166]: Same comments for the else (no edge weight) path.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+               static_cast<float>(src_prop),
+               static_cast<float>(dst_prop));
+        return dst_prop;
+      },
+      result_t{0},
+      cugraph::reduce_op::plus<result_t>{},
+      outputs.begin());
+
+    auto outputs_title                 = std::string("outputs_").append(std::to_string(comm_rank));
+    size_t max_nr_of_elements_to_print = 10;
```
Comment (seunghwak) [inline cpp/examples/advanced/cugraph_operations/graph_operations.cu:220]: And considering that we are printing for every edge in the above device functor, should we really limit the elements to print here?

File: cpp/examples/advanced/cugraph_operations/CMakeLists.txt
[the remark this answers is not in the document]
Comment (seunghwak) [reply cpp/examples/advanced/cugraph_operations/CMakeLists.txt:1]: Or better say graph_primitives as this file is really about using per_v_transform_reduce_incoming_e.

File: cpp/examples/advanced/graph_partition/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+    "../../../src"
+)
+set_target_properties(csvfileutil PROPERTIES CUDA_ARCHITECTURES "native")
+target_link_libraries(csvfileutil PUBLIC cugraph::cugraph NCCL::NCCL MPI::MPI_CXX)
+
+# Configure your project here
+add_executable(vertex_and_edge_partition vertex_and_edge_partition.cu)
+set_target_properties(vertex_and_edge_partition PROPERTIES CUDA_ARCHITECTURES "native")
+target_link_libraries(vertex_and_edge_partition PRIVATE csvfileutil)
+target_compile_features(vertex_and_edge_partition PRIVATE cxx_std_17)
```
Comment (seunghwak) [inline cpp/examples/advanced/graph_partition/CMakeLists.txt:45]: We are mixing "graph_partition" and "vertex_and_edge_partition", and better be consistent.

File: cpp/examples/advanced/graph_partition/vertex_and_edge_partition.cu [hunk 18]
```diff
@@ -0,0 +1,365 @@
...
+
+void initialize_mpi_and_set_device(int argc, char** argv)
+{
+  RAFT_MPI_TRY(MPI_Init(&argc, &argv));
+
+  int comm_rank{};
+  RAFT_MPI_TRY(MPI_Comm_rank(MPI_COMM_WORLD, &comm_rank));
+
+  int comm_size{};
+  RAFT_MPI_TRY(MPI_Comm_size(MPI_COMM_WORLD, &comm_size));
```
Comment (seunghwak) [inline cpp/examples/advanced/graph_partition/vertex_and_edge_partition.cu:41]: Where are we using comm_size?

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+
+void initialize_mpi_and_set_device(int argc, char** argv)
+{
+  RAFT_MPI_TRY(MPI_Init(&argc, &argv));
+
+  int comm_rank{};
+  RAFT_MPI_TRY(MPI_Comm_rank(MPI_COMM_WORLD, &comm_rank));
+
+  int comm_size{};
+  RAFT_MPI_TRY(MPI_Comm_size(MPI_COMM_WORLD, &comm_size));
```
Comment (seunghwak) [inline cpp/examples/advanced/cugraph_operations/graph_operations.cu:47]: Where are we using comm_size?

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+
+std::unique_ptr<raft::handle_t> initialize_mg_handle(std::string const& allocation_mode = "cuda")
+{
+  int comm_rank{};
+  RAFT_MPI_TRY(MPI_Comm_rank(MPI_COMM_WORLD, &comm_rank));
+
+  std::set<std::string> possible_allocation_modes = {"cuda", "pool", "binning", "managed"};
+
+  if (possible_allocation_modes.find(allocation_mode) == possible_allocation_modes.end()) {
+    if (!comm_rank) {
```
Comment (seunghwak) [inline cpp/examples/advanced/cugraph_operations/graph_operations.cu:62]: comm_rank == 0 is more intuitive. Here, the intention is to print in only the root process. I assume "comm_rank == 0" much better conveys this intention than !comm_rank

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 21]
```diff
@@ -0,0 +1,253 @@
...
+      std::for_each(possible_allocation_modes.cbegin(),
+                    possible_allocation_modes.cend(),
+                    [](std::string mode) { std::cout << mode << std::endl; });
+    }
+    RAFT_MPI_TRY(MPI_Finalize());
+
+    exit(0);
+  }
+
+  if (!comm_rank) {
```
Comment (seunghwak) [inline cpp/examples/advanced/cugraph_operations/graph_operations.cu:75]: Same here and in this function in other example files as well.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 22]
```diff
@@ -0,0 +1,253 @@
...
+    std::cout << "Using '" << allocation_mode
+              << "' allocation mode to create device memory resources." << std::endl;
+  }
+  std::shared_ptr<rmm::mr::device_memory_resource> resource =
+    cugraph::test::create_memory_resource(allocation_mode);
```
Comment (seunghwak) [inline cpp/examples/advanced/cugraph_operations/graph_operations.cu:80]: I am not sure using a test utility function is a right thing to do in the example code.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 23]
```diff
@@ -0,0 +1,253 @@
...
+
+/**
+ * @brief This function reads graph from an input csv file and
+ * display vertex and edge partitions.
+ */
+
+template <typename vertex_t, typename edge_t, typename weight_t, bool multi_gpu>
+void perform_example_graph_operations(raft::handle_t const& handle,
+                                      std::string const& csv_graph_file_path,
+                                      const bool weighted = false)
```
[the remark this answers is not in the document]
Comment (seunghwak) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu:108]: And you aren't using const in pass-by-value input parameter elsewhere (e.g. https://github.com/rapidsai/cugraph/pull/4128/files#diff-b59fa50933677b1fb19604d2da9f9adf1fe13dc5dffb2436953b2815a2da2415R102).

And there exists some debate about using const in pass-by-value input parameters (e.g. https://stackoverflow.com/questions/117293/does-using-const-on-function-parameters-have-any-effect-why-does-it-not-affect) but we are not using const in pass-by-value parameter in our API. So, I think we'd better be consistent.

File: cpp/examples/simple/multi_gpu_application/CMakeLists.txt [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+include(../../../cmake/thirdparty/get_nccl.cmake)
+
+add_library(csvfileutil STATIC ../../../tests/utilities/csv_file_utilities.cu)
+target_include_directories(csvfileutil
+  PUBLIC
+    "../../../../thirdparty/mmio"
+    "../../../tests"
+  PRIVATE
+    "../../../include"
+    "../../../src"
```
Comment (seunghwak) [inline cpp/examples/simple/multi_gpu_application/CMakeLists.txt:37]: libcugraph users (not developers) should not rely on files under the src directory.

File: cpp/examples/README.md
Comment (acostadon) [inline cpp/examples/README.md:1]: a link or explanation of the memory allocation types would help

Comment (naimnv) [reply cpp/examples/README.md:1]: To simplify all the example I remove memory allocation  parameter and used cuda for all of the examples.

File: cpp/examples/README.md [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+
+```sh
+~/cugraph/cpp/examples$./build.sh
+```
+
+## Run instructions
+
+For single_gpu
+
+`path_to_executable path_to_a_csv_graph_file [memory allocation mode]`
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/README.md:30]: As we are only focusing on example rather than real application, removed the memory allocation mode parameter and used rmm::mr::cuda_memory_resource for all the examples.

File: cpp/examples/advanced/cugraph_operations/CMakeLists.txt
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/CMakeLists.txt:1]: Used graph_operations.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu (hunk 23 above)
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: This parameter is not used anymore.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 26]
```diff
@@ -0,0 +1,253 @@
...
+  std::cout << "Rank_" << comm_rank << ", reading graph from " << csv_graph_file_path << std::endl;
+
+  bool renumber = true;  // must be true for distributed graph.
+
+  // Read a graph (along with edge properties e.g. edge weights, if provided) from
+  // the input csv file
+
+  auto [graph, edge_weights, renumber_map] =
+    cugraph::test::read_graph_from_csv_file<vertex_t, edge_t, weight_t, false, multi_gpu>(
+      handle, csv_graph_file_path, weighted, renumber);
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: Used edge list.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 27]
```diff
@@ -0,0 +1,253 @@
...
+  }
+
+  std::string const& csv_graph_file_path = argv[1];
+  std::string const& allocation_mode     = argc < 3 ? "cuda" : argv[2];
+
+  initialize_mpi_and_set_device(argc, argv);
+  std::unique_ptr<raft::handle_t> handle = initialize_mg_handle(allocation_mode);
+
+  auto const comm_rank = handle->get_comms().get_rank();
+  auto const comm_size = handle->get_comms().get_size();
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: removed.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 28]
```diff
@@ -0,0 +1,253 @@
...
+void perform_example_graph_operations(raft::handle_t const& handle,
+                                      std::string const& csv_graph_file_path,
+                                      const bool weighted = false)
+{
+  auto const comm_rank = handle.get_comms().get_rank();
+  auto const comm_size = handle.get_comms().get_size();
+
+  std::cout << "Rank_" << comm_rank << ", reading graph from " << csv_graph_file_path << std::endl;
+
+  bool renumber = true;  // must be true for distributed graph.
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: Changed.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 29]
```diff
@@ -0,0 +1,253 @@
...
+  // the input csv file
+
+  auto [graph, edge_weights, renumber_map] =
+    cugraph::test::read_graph_from_csv_file<vertex_t, edge_t, weight_t, false, multi_gpu>(
+      handle, csv_graph_file_path, weighted, renumber);
+
+  // Non-owning view of the graph object
+  auto graph_view = graph.view();
+  // Number of vertices mapped to this process, ie the size of
+  // the vertex partition assigned to this process
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: Mentioned one-process-per-GPU

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 30]
```diff
@@ -0,0 +1,253 @@
...
+
+  std::cout << "Rank_" << comm_rank << ", reading graph from " << csv_graph_file_path << std::endl;
+
+  bool renumber = true;  // must be true for distributed graph.
+
+  // Read a graph (along with edge properties e.g. edge weights, if provided) from
+  // the input csv file
+
+  auto [graph, edge_weights, renumber_map] =
+    cugraph::test::read_graph_from_csv_file<vertex_t, edge_t, weight_t, false, multi_gpu>(
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: defined

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 31]
```diff
@@ -0,0 +1,253 @@
...
+
+  // Non-owning view of the graph object
+  auto graph_view = graph.view();
+  // Number of vertices mapped to this process, ie the size of
+  // the vertex partition assigned to this process
+
+  vertex_t size_of_the_vertex_partition_assigned_to_this_process =
+    graph_view.local_vertex_partition_range_size();
+
+  // Non-owning of the edge edge_weights object
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: Changed.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 32]
```diff
@@ -0,0 +1,253 @@
...
+  auto edge_weight_view = edge_weights ? std::make_optional((*edge_weights).view()) : std::nullopt;
+
+  using graph_view_t = cugraph::graph_view_t<vertex_t, edge_t, false, multi_gpu>;
+
+  //
+  // As an example operation, compute the weighted average of the properties of
+  // neighboring vertices, weighted by the edge weights, if the input graph is weighted;
+  // Otherwise, compute the simple average.
+  //
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: Tried to be consistent.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 33]
```diff
@@ -0,0 +1,253 @@
...
+    using result_t      = weight_t;
+    auto vertex_weights = compute_out_weight_sums(handle, graph_view, *edge_weight_view);
+
+    cugraph::edge_src_property_t<graph_view_t, result_t> src_vertex_weights_cache(handle,
+                                                                                  graph_view);
+
+    cugraph::edge_dst_property_t<graph_view_t, result_t> dst_vertex_weights_cache(handle,
+                                                                                  graph_view);
+
+    update_edge_src_property(handle, graph_view, vertex_weights.begin(), src_vertex_weights_cache);
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: Resolved.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 34]
```diff
@@ -0,0 +1,253 @@
...
+
+    update_edge_dst_property(handle, graph_view, vertex_weights.begin(), dst_vertex_weights_cache);
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: added cugraph::

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 35]
```diff
@@ -0,0 +1,253 @@
...
+
+    rmm::device_uvector<result_t> outputs(size_of_the_vertex_partition_assigned_to_this_process,
+                                          handle.get_stream());
+
+    per_v_transform_reduce_incoming_e(
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: added cugraph::

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 36]
```diff
@@ -0,0 +1,253 @@
...
+  using graph_view_t = cugraph::graph_view_t<vertex_t, edge_t, false, multi_gpu>;
+
+  //
+  // As an example operation, compute the weighted average of the properties of
+  // neighboring vertices, weighted by the edge weights, if the input graph is weighted;
+  // Otherwise, compute the simple average.
+  //
+  if (weighted) {
+    using result_t      = weight_t;
+    auto vertex_weights = compute_out_weight_sums(handle, graph_view, *edge_weight_view);
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: Added cugraph::

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 37]
```diff
@@ -0,0 +1,253 @@
...
+  } else {
+    using result_t      = edge_t;
+    auto vertex_weights = graph_view.compute_out_degrees(handle);
+
+    cugraph::edge_src_property_t<graph_view_t, result_t> src_vertex_weights_cache(handle,
+                                                                                  graph_view);
+    cugraph::edge_dst_property_t<graph_view_t, result_t> dst_vertex_weights_cache(handle,
+                                                                                  graph_view);
+
+    update_edge_src_property(handle, graph_view, vertex_weights.begin(), src_vertex_weights_cache);
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: added cugraph::

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 38]
```diff
@@ -0,0 +1,253 @@
...
+
+    update_edge_dst_property(handle, graph_view, vertex_weights.begin(), dst_vertex_weights_cache);
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: added cugraph::

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu (hunk 14 above)
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: Changed.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu (hunk 15 above)
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: changed.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu (hunk 16 above)
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: Removed. Not needed anymore.

File: cpp/examples/advanced/graph_partition/CMakeLists.txt (hunk 17 above)
Comment (naimnv) [reply cpp/examples/advanced/graph_partition/CMakeLists.txt]: Changed.

File: cpp/examples/advanced/graph_partition/vertex_and_edge_partition.cu (hunk 18 above)
Comment (naimnv) [reply cpp/examples/advanced/graph_partition/vertex_and_edge_partition.cu]: Removed.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu (hunk 19 above)
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: removed.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu (hunk 20 above)
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: Changed.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu (hunk 21 above)
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: changed.

File: cpp/examples/simple/multi_gpu_application/CMakeLists.txt (hunk 24 above)
Comment (naimnv) [reply cpp/examples/simple/multi_gpu_application/CMakeLists.txt]: Removed.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu (hunk 22 above)
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: Removed.

File: cpp/examples/advanced/cugraph_operations/graph_operations.cu [hunk 39]
```diff
@@ -0,0 +1,253 @@
...
+  // Number of vertices mapped to this process, ie the size of
+  // the vertex partition assigned to this process
+
+  vertex_t size_of_the_vertex_partition_assigned_to_this_process =
+    graph_view.local_vertex_partition_range_size();
+
+  // Non-owning of the edge edge_weights object
+  auto edge_weight_view = edge_weights ? std::make_optional((*edge_weights).view()) : std::nullopt;
+
+  using graph_view_t = cugraph::graph_view_t<vertex_t, edge_t, false, multi_gpu>;
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/advanced/cugraph_operations/graph_operations.cu]: It didn't work, perhaps need to remove const qualifier.

File: cpp/examples/README.md [hunk 40]
```diff
@@ -0,0 +1,43 @@
...
+
+  -  cugraph_operations: example code for using cugraph primitives for simple graph operations needed to implement graph algorithms. 
+
+## Build instructions:
+
+Run `build.sh` to from examples dir to build the above listed examples.
+
+```sh
+~/cugraph/cpp/examples$./build.sh
+```
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/README.md:24]: As these examples are only to show/explain how to use cugraph, I'm not sure if it's necessary

File: cpp/examples/README.md (hunk 12 above)
Comment (naimnv) [reply cpp/examples/README.md]: changed.

File: cpp/examples/users/multi_gpu_application/mg_graph_algorithms.cpp [hunk 41]
```diff
@@ -0,0 +1,292 @@
...
+{
+  size_t num_edges = edge_srcs.size();
+  assert(edge_dsts.size() == num_edges);
+  if (edge_wgts.has_value()) { assert((*edge_wgts).size() == num_edges); }
+
+  auto const comm_rank = handle.get_comms().get_rank();
+  auto const comm_size = handle.get_comms().get_size();
+
+  //
+  // Assign part of the edge list to each GPU
```
Comment (acostadon) [inline cpp/examples/users/multi_gpu_application/mg_graph_algorithms.cpp:95]: a little explanation could help here.

File: cpp/examples/users/multi_gpu_application/mg_graph_algorithms.cpp [hunk 42]
```diff
@@ -0,0 +1,292 @@
...
+
+  raft::update_device(d_edge_srcs.data(), edge_srcs.data() + start, work_size, handle.get_stream());
+  raft::update_device(d_edge_dsts.data(), edge_dsts.data() + start, work_size, handle.get_stream());
+  if (d_edge_wgts.has_value()) {
+    raft::update_device(
+      (*d_edge_wgts).data(), (*edge_wgts).data() + start, work_size, handle.get_stream());
+  }
+
+  //
+  // Shuffle edges to proper GPU
```
Comment (acostadon) [inline cpp/examples/users/multi_gpu_application/mg_graph_algorithms.cpp:118]: a little explanation could help here.

File: cpp/examples/developers/graph_operations/CMakeLists.txt
Comment (acostadon) [inline cpp/examples/developers/graph_operations/CMakeLists.txt:1]: is the user expected to know what the settings here do?

Comment (naimnv) [reply cpp/examples/developers/graph_operations/CMakeLists.txt:1]: Not really. Mainly for the easy of building all the examples. User can build all the examples in one go just running ./build.sh, which in turn use each CMakeLists.txt
It's documented in README.

**Review by ChuckHastings — APPROVED** [review]
LTGM, code wise.  Agree with Don's documentation comments... will wait to merge until those are addressed.

File: cpp/examples/users/multi_gpu_application/mg_graph_algorithms.cpp (hunk 42 above)
Comment (naimnv) [reply cpp/examples/users/multi_gpu_application/mg_graph_algorithms.cpp]: elaborated.

File: cpp/examples/users/multi_gpu_application/mg_graph_algorithms.cpp (hunk 41 above)
Comment (naimnv) [reply cpp/examples/users/multi_gpu_application/mg_graph_algorithms.cpp]: Added explanation.

File: cpp/examples/users/multi_gpu_application/mg_graph_algorithms.cpp [hunk 43]
```diff
@@ -0,0 +1,292 @@
...
+    --gpu_row_comm_size;
+  }
+
+  cugraph::partition_manager::init_subcomm(*handle, gpu_row_comm_size);
+
+  return std::move(handle);
+}
+
+/**
+ * @brief Create a graph from edge sources, destinitions, and optional weights
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/examples/users/multi_gpu_application/mg_graph_algorithms.cpp]: Fixed.

**Review by acostadon — APPROVED** [review]
updates look great

**Review by seunghwak — APPROVED** [review]
LGTM
